core/dbt/contracts/util.py

```diff
@@ -69,22 +70,41 @@ class AdditionalPropertiesMixin:
     """
     ADDITIONAL_PROPERTIES = True
 
+    # This takes attributes in the dictionary that are
+    # not in the class definitions and puts them in an
+    # _extra dict in the class
     @classmethod
-    def from_dict(cls, data, validate=True):
-        self = super().from_dict(data=data, validate=validate)
-        keys = self.to_dict(validate=False, omit_none=False)
+    def __pre_deserialize__(cls, data, options=None):
+        # dir() did not work because fields with
+        # metadata settings are not found
+        # The original version of this would create the
+        # object first and then update extra with the
+        # extra keys, but that won't work here, so
+        # we're copying the dict so we don't insert the
+        # _extra in the original data. This also requires
+        # that Mashumaro actually build the '_extra' field
+        cls_keys = cls._get_field_names()
+        new_dict = {}
         for key, value in data.items():
-            if key not in keys:
-                self.extra[key] = value
-        return self
+            if key not in cls_keys and key != '_extra':
+                if '_extra' not in new_dict:
+                    new_dict['_extra'] = {}
+                new_dict['_extra'][key] = value
+            else:
```

Comment:
should this read:
```
elif key != '_extra':
```

or is that not a concern here? Just trying to follow the logic in the conditional here and might be missing some nuance

Reply:
We're putting any "extra" keys that are not in the class itself into an '_extra' dict, so that line creates the '_extra' key if it doesn't exist. That particular area was a big headache. Would be nice to make it nicer, but 'elif' won't work there. A previous broken thing that happened was deeply nested _extra keys in _extra dicts.

Reply:
ok.... sounds fun... thanks for the pointer :)